### PR TITLE
Support 0-sized batches in SyncBatchNorm cuda ops

### LINKED ISSUE #78656: Allow batch_norm_backward_elemt and batch_norm_gather_stats_with_counts handle 0 counts
### 🚀 The feature, motivation and pitch

#36530 requested empty batch support for SyncBatchNorm, which is later added in #74944. However, the fix in #74944 broke CUDA graph capturing as the new code introduced sync on CPU. See #78549.

For v1.12, the temporary fix is to guard the sync in new code with [is_current_stream_capturing](https://pytorch.org/docs/master/generated/torch.cuda.is_current_stream_capturing.html?highlight=is_current_stream_capturing#torch.cuda.is_current_stream_capturing). A better long-term fix would be updating the CUDA kernels for `batch_norm_backward_elemt` and `batch_norm_gather_stats_with_counts` to support zero counts, so that empty batch can work with CUDA graph capturing as well. 

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by nlgranger**
Fixes #78656

~~Haven't been able to build and therefore test this locally, hence the WIP status.~~

### CODE REVIEW

**Comment by nlgranger** [comment]
There are no tests for the case of 0-sized batches, I'll try to learn how to write one but pointers and recommendations are welcome!

File: test/test_nn.py [hunk 1]
```diff
@@ -6869,6 +6869,56 @@ def test_convert_sync_batchnorm(self):
...
+                running_var_,
+                momentum,
+                eps,
+                count.to(torch.get_default_dtype())
+            )
+
+            self.assertEqual(mean_, mean_batch)
+            self.assertEqual(inv_std_, inv_std_batch)
+            self.assertEqual(running_mean_, running_mean_batch)
+            self.assertEqual(running_var_, running_var_batch, atol=1.3e-6, rtol=1e-5)
```
Comment (nlgranger) [inline test/test_nn.py:6920]: Assert fails with default precision target, it looks like a precision issue.
Alternately this could be the c expected value which has an error.
Or my modification messed something up, but I find it unlikely that running variance would be the sole victim.

**Comment by nlgranger** [comment]
Can I retrieve the built wheel from the CI? I'd like to run a few tests locally but I failed to build pytorch on my computer.

**Review by lawchingman — APPROVED** [review]

**Review by ngimel — APPROVED** [review]

**Comment by kit1980** [comment]
Is this ready to me merged?

**Comment by nlgranger** [comment]
I'm done with the code but I need reviews on:
- Code quality
- Test coverage
- The test which only passes with lower precision (see above)

**Comment by nlgranger** [comment]
Rebased on main, can anyone remove the stale label please?

**Comment by nlgranger** [comment]
@albanD @jbschlosser @mikaylagawarecki sorry for bumping but could you please confirm if you are available for reviewing? Otherwise this PR will remain stuck.

**Comment by mikaylagawarecki** [comment]
@nlgranger Apologies for the delay, I will take a look at this later today

**Review by mikaylagawarecki — APPROVED** [review]
Thanks and apologies for the delay on this! I think this is ok and does not break the existing distributed tests that check for these cases [here](https://github.com/pytorch/pytorch/blob/main/test/distributed/test_c10d_common.py#L862-L948)

File: test/test_nn.py [hunk 2]
```diff
@@ -7134,6 +7134,67 @@ def test_convert_sync_batchnorm(self):
...
+                eps,
+                count.to(torch.get_default_dtype())
+            )
+
+            if n > 0:
+                self.assertEqual(mean_pred, mean_expected)
+                self.assertEqual(inv_std_pred, inv_std_expected)
+
+            self.assertEqual(running_mean_pred, running_mean_expected)
+            self.assertEqual(running_var_pred, running_var_expected, atol=1.3e-6, rtol=1e-5)
```
Comment (mikaylagawarecki) [inline test/test_nn.py:7196]: ah, I was perplexed by why the outputs are the same regardless of the input i, but I now see this :)

https://github.com/pytorch/pytorch/blob/2039f30c062ec225bf1add9a887c7d8b30a8b6a9/aten/src/ATen/native/cuda/Normalization.cu#L614
